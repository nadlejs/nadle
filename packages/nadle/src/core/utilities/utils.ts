import Path from "node:path";

import c from "tinyrainbow";

import { DOT } from "./constants.js";

export function noop() {}

export function capitalize(str: string): string {
	if (str.length === 0) {
		return str;
	}

	return str.charAt(0).toUpperCase() + str.slice(1);
}

const MILLISECONDS_PER_SECOND = 1000;
const MILLISECONDS_PER_MINUTE = 60 * MILLISECONDS_PER_SECOND;
const MILLISECONDS_PER_HOUR = 60 * MILLISECONDS_PER_MINUTE;

export function formatTime(ms: number): string {
	const hours = Math.floor(ms / MILLISECONDS_PER_HOUR);
	const minutes = Math.floor((ms % MILLISECONDS_PER_HOUR) / MILLISECONDS_PER_MINUTE);
	const seconds = Math.floor((ms % MILLISECONDS_PER_MINUTE) / MILLISECONDS_PER_SECOND);
	const milliseconds = Math.floor(ms % MILLISECONDS_PER_SECOND);

	if (hours > 0) {
		if (minutes > 0) {
			return `${hours}h${minutes}m`;
		}

		return `${hours}h`;
	}

	if (minutes > 0) {
		if (seconds > 0) {
			return `${minutes}m${seconds}s`;
		}

		return `${minutes}m`;
	}

	if (seconds > 0) {
		return `${seconds}s`;
	}

	return `${milliseconds}ms`;
}

export function normalizeGlobPath(path: string) {
	if (path.startsWith(DOT)) {
		return path;
	}

	return `.${Path.sep}${path}`;
}

export function clamp(value: number, min: number, max: number): number {
	return Math.min(Math.max(value, min), max);
}

export function bindObject<T, K extends keyof T>(obj: T, keys: K[]): { [P in K]: T[P] extends (...args: any[]) => any ? T[P] : never } {
	return Object.fromEntries(
		keys.map((key) => {
			const value = obj[key];

			if (typeof value !== "function") {
				throw new Error(`Property ${String(key)} is not a function`);
			}

			return [key, value.bind(obj)];
		})
	) as any;
}

export function highlight(text: string): string {
	return c.bold(c.yellow(text));
}
