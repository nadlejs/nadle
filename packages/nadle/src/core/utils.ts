export function capitalize(str: string): string {
	if (str.length === 0) {
		return str;
	}

	return str.charAt(0).toUpperCase() + str.slice(1);
}

export function formatTime(time: number): string {
	if (time > 1000) {
		return `${(time / 1000).toFixed(2)}s`;
	}

	return `${Math.round(time)}ms`;
}

export function formatTimeString(date: Date): string {
	return date.toTimeString().split(" ")[0];
}
