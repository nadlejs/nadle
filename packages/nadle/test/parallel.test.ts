import { it, describe } from "vitest";

import { createExec, expectFail, expectPass } from "./utils.js";

describe("Parallel", () => {
	const exec = createExec({ config: "parallel" });

	it("should run tasks parallel", async () => {
		await expectPass(exec`--max-workers 2 task-A`, [{ pattern: /task-A\.[01]/g, replacement: () => "task-A.[01]" }]);
	}, 8000);

	it("should stop on error immediately", async () => {
		await expectFail(() => exec`--max-workers 2 X`, [{ pattern: /task-A\.[12]/g, replacement: () => "task-A.[12]" }]);
	}, 4000);
});
