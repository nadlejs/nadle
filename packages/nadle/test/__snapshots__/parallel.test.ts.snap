// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parallel > should run tasks parallel > stdout 1`] = `
[log] <Bold><Cyan>🛠️ Welcome to Nadle {version}!</Cyan></BoldDim>
[log] <Dim>Loading configuration file from: /ROOT/test/__fixtures__/main/nadle.parallel.ts</BoldDim>
[log] <Dim>Using 2 workers for task execution</BoldDim>
[log] <Yellow>></Yellow> Task <Bold>task-A.[01]</BoldDim> started

[log] <Yellow>></Yellow> Task <Bold>task-A.[01]</BoldDim> started

[log] 
<Green>✓</Green> Task <Bold>task-A.[01]</BoldDim> done in {duration}
[log] <Yellow>></Yellow> Task <Bold>task-A.2</BoldDim> started

[log] 
<Green>✓</Green> Task <Bold>task-A.[01]</BoldDim> done in {duration}
[log] 
<Green>✓</Green> Task <Bold>task-A.2</BoldDim> done in {duration}
[log] <Yellow>></Yellow> Task <Bold>task-A</BoldDim> started

[log] 
<Green>✓</Green> Task <Bold>task-A</BoldDim> done in {duration}
[log] 
<Bold><Green>RUN SUCCESSFUL</Green></BoldDim> in <Bold>{duration}</BoldDim> <Dim>(<Bold>4</BoldDim><Dim> tasks executed)</BoldDim>
`;

exports[`parallel > should stop on error immediately > stderr 1`] = ``;

exports[`parallel > should stop on error immediately > stdout 1`] = `
[log] <Bold><Cyan>🛠️ Welcome to Nadle {version}!</Cyan></BoldDim>
[log] <Dim>Loading configuration file from: /ROOT/test/__fixtures__/main/nadle.parallel.ts</BoldDim>
[log] <Dim>Using 2 workers for task execution</BoldDim>
[log] <Yellow>></Yellow> Task <Bold>task-A.[12]</BoldDim> started

[log] <Yellow>></Yellow> Task <Bold>task-A.[12]</BoldDim> started

[log] 
<Green>✓</Green> Task <Bold>task-A.[12]</BoldDim> done in {duration}
[log] 
<Green>✓</Green> Task <Bold>task-A.[12]</BoldDim> done in {duration}
[log] <Yellow>></Yellow> Task <Bold>throwable</BoldDim> started

[log] 
<Red>×</Red> Task <Bold>throwable</BoldDim> failed in {duration}
[log] 
<Bold><Red>RUN FAILED</Red></BoldDim> in <Bold>{duration}</BoldDim> <Dim>(<Bold>2</BoldDim><Dim> tasks executed, <Bold>1</BoldDim><Dim> task failed)</BoldDim>
[log] 
For more details, re-run the command with the <Yellow>--stacktrace</Yellow> option to display the full error and help identify the root cause.
`;
